controlers/contacts/index.js

```diff
@@ -7,17 +8,113 @@ const {
   deleteContact,
   updateFavorite,
 } = require("./services");
+const Users = require("../../models/users-models");
+const bcrypt = require("bcrypt");
 
 const schema = Joi.object({
   name: Joi.string().min(3).max(20).required(),
   email: Joi.string().email({ minDomainSegments: 2 }).required(),
   phone: Joi.string().min(9).required(),
   favorite: Joi.boolean,
 });
+const schemaRegister = Joi.object({
+  email: Joi.string().email({ minDomainSegments: 2 }).required(),
+  password: Joi.string().required(),
+  subscription: Joi.string(),
+});
 const schemaFav = Joi.object({
   favorite: Joi.boolean().required(),
 });
 
+const currentUser = async (req, res) => {
+  const user = req.user;
+  if (user) {
+    return res
+      .status(200)
+      .json({ email: user.email, subscription: user.subscription });
+  } else {
+    return res.status(401).json({ message: "Not authorized" });
+  }
+};
+
+const userSignup = async (req, res, next) => {
+  const { email, password, subscription } = req.body;
+  const { error } = schemaRegister.validate(req.body);
+  if (error) {
+    return res.status(400).send(error.details[0].message);
+  }
+  try {
+    const user = await Users.findOne({ email });
+    if (user) {
+      res.status(409).json({ message: "Email in use" });
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    const newUser = new Users({
+      email,
+      password: hashedPassword,
+      subscription,
+    });
+
+    await newUser.save();
+    return res.status(201).json({
+      user: {
+        email: newUser.email,
+        subscription: newUser.subscription,
+      },
+    });
+  } catch (err) {
+    next(err);
+  }
+};
+
+const userLogin = async (req, res) => {
+  const { email, password } = req.body;
+  const { error } = schemaRegister.validate(req.body);
+  if (error) {
+    return res.status(400).send(error.details[0].message);
+  }
+  try {
+    const user = await Users.findOne({ email });
+    if (!user) {
+      return res.status(401).json({ message: "Email or password is wrong" });
+    }
+    const isPasswordCorrect = await user.validatePassword(password);
+    if (isPasswordCorrect) {
+      const payload = {
+        id: user._id,
+      };
+      const token = jwt.sign(payload, process.env.AUTH_SECRET, {
+        expiresIn: "12h",
+      });
+      user.token = token;
+      await user.save();
+      return res.status(200).json({
+        token,
+        user: { email: user.email, subscription: user.subscription },
+      });
+    } else {
+      return res.status(401).json({ message: "Wrong password" });
+    }
+  } catch (err) {
+    return res.status(500).json({ message: "Internal server error" });
+  }
+};
+const userLogout = async (req, res, next) => {
+  try {
+    const userId = req.user._id;
+    const user = await Users.findById(userId);
+    if (!user) {
+      return res.status(401).json({ message: "Not authorized" });
```

Comment:
authMiddleware jus sprawdzilo czy jest authorized czy nie, czyli jesli tu jestesmy to user zostal pomyslnie uwierzytelniony